<!DOCTYPE html>
<html lang="es">
<head>
    <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Lanzamiento de Excepciones</title>

    <link rel="alternate" href="https://campusempresa.com/python/raising-exceptions" hreflang="es" />
	<link rel="alternate" href="https://campusempresa.cat/python/raising-exceptions" hreflang="ca" />
	<link rel="alternate" href="https://enterprisecampus.net/python/raising-exceptions" hreflang="en" />
    
	<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.3/dist/css/bootstrap.min.css" rel="stylesheet">
	<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap-icons/font/bootstrap-icons.css">
	<link href="/css/site.css" rel="stylesheet">
	
    <script src="https://code.jquery.com/jquery-3.7.1.min.js"></script>
  	<script type="text/javascript" src="js/math_init.js"></script>
  	<script type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/startup.js"></script>
  	<script type="text/javascript" src="/js/cookie.js"></script>
  	<script type="text/javascript" src="/js/main.js"></script>
</head>

<body>
    <div id="content">
		<div id="header" class="container-xxl">
	<div class="row">
		<div class="col-8 p-0">
			<h1 class="m-0 p-0">
				<a href="/"><img src="/img/logo_header.png" style="visibility:hiddenxx;"></a>
			</h1>
		</div>
		<div class="col-4 p-0 text-end">
			<h2 id="main_title"><cite>Construyendo la sociedad de hoy y del mañana</cite></h2>
			<h3 id="main_subtitle"></h3>
		</div>
	</div>
</div>
<div class="container-xxl" style="margin-top: -1em;">
	<div class="row">
		<div class="col-12 p-0 m-0 text-end">
										<a href="https://enterprisecampus.net/python/raising-exceptions" class="px-2">EN</a></b>
				|
				<b class="px-2">ES</b>
				|
				<a href="https://campusempresa.cat/python/raising-exceptions" class="px-2">CA</a>
								</div>
	</div>
</div>
   <div class="top-bar container-fluid">
	<div class="container-xxl">
		<div class="row">
			<div class="col" id="left_menu">
				<a href="/objective">El Proyecto</a>
				<a href="/about">Sobre nosotros</a>
				<a href="/contribute">Contribuir</a>
				<a href="/donate">Donaciones</a>
				<a href="/licence">Licencia</a>
			</div>
		</div>
	</div>
   </div>

<div class="container-xxl" id="main_content">
	<div class="row">
		<div class="col-12 col-lg-8">
			<div id="nav1" class="navigation"></div>
			<div id="inner_content"><div class='row navigation'>
	<div class='col-4'>
					<a href='handling-exceptions'>&#x25C4;Manejo de Excepciones</a>
			</div>
	<div class='col-4 text-center'>
		<a href="./" class="title">Lanzamiento de Excepciones</a>
	</div>
	<div class='col-4 text-end'>
					<a href='custom-exceptions'>Excepciones Personalizadas &#x25BA;</a>
			</div>
</div>
<div class='content'></div><h1>Introducción</h1>
<div class='content'><p>En Python, las excepciones son una forma de manejar errores que ocurren durante la ejecución de un programa. Lanzar excepciones te permite señalar que ha ocurrido un error y proporcionar información sobre la naturaleza del error. Esta sección cubrirá los conceptos básicos del lanzamiento de excepciones, cómo crear excepciones personalizadas y técnicas avanzadas para manejar excepciones.</p>
</div><h1>Conceptos Clave</h1>
<div class='content'><ul>
<li><strong>Excepción</strong>: Un error que ocurre durante la ejecución de un programa.</li>
<li><strong>Lanzar una Excepción</strong>: El acto de señalar que ha ocurrido un error usando la declaración <code>raise</code>.</li>
<li><strong>Excepciones Personalizadas</strong>: Excepciones definidas por el usuario que extienden la clase base <code>Exception</code>.</li>
</ul>
</div><h1>Lanzamiento de Excepciones</h1>
<div class='content'></div><h2>Sintaxis Básica</h2>
<div class='content'><p>Para lanzar una excepción en Python, usas la declaración <code>raise</code> seguida de una instancia de una clase de excepción.</p>
</div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("IyBFamVtcGxvOiBMYW56YW5kbyB1bmEgZXhjZXBjacOzbiBiw6FzaWNhCnJhaXNlIEV4Y2VwdGlvbigiRXN0YSBlcyB1bmEgZXhjZXBjacOzbiBnZW7DqXJpY2EiKQ=="))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'># Ejemplo: Lanzando una excepci&oacute;n b&aacute;sica
raise Exception(&quot;Esta es una excepci&oacute;n gen&eacute;rica&quot;)</pre></div><div class='content'></div><h2>Excepciones Comunes Incorporadas</h2>
<div class='content'><p>Python proporciona varias excepciones incorporadas que puedes usar para señalar diferentes tipos de errores:</p>
<ul>
<li><code>ValueError</code>: Se lanza cuando una función recibe un argumento del tipo correcto pero de valor inapropiado.</li>
<li><code>TypeError</code>: Se lanza cuando una operación o función se aplica a un objeto de tipo inapropiado.</li>
<li><code>IndexError</code>: Se lanza cuando un subíndice de secuencia está fuera de rango.</li>
<li><code>KeyError</code>: Se lanza cuando una clave de diccionario no se encuentra.</li>
</ul>
</div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("IyBFamVtcGxvOiBMYW56YW5kbyB1biBWYWx1ZUVycm9yCmRlZiBkaXZpZGUoYSwgYik6CiAgICBpZiBiID09IDA6CiAgICAgICAgcmFpc2UgVmFsdWVFcnJvcigiTm8gc2UgcHVlZGUgZGl2aWRpciBwb3IgY2VybyIpCiAgICByZXR1cm4gYSAvIGIKCnRyeToKICAgIHJlc3VsdCA9IGRpdmlkZSgxMCwgMCkKZXhjZXB0IFZhbHVlRXJyb3IgYXMgZToKICAgIHByaW50KGUp"))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'># Ejemplo: Lanzando un ValueError
def divide(a, b):
    if b == 0:
        raise ValueError(&quot;No se puede dividir por cero&quot;)
    return a / b

try:
    result = divide(10, 0)
except ValueError as e:
    print(e)</pre></div><div class='content'></div><h2>Creación de Excepciones Personalizadas</h2>
<div class='content'><p>Puedes crear tus propias excepciones definiendo una nueva clase que herede de la clase <code>Exception</code>.</p>
</div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("IyBFamVtcGxvOiBDcmVhbmRvIHVuYSBleGNlcGNpw7NuIHBlcnNvbmFsaXphZGEKY2xhc3MgQ3VzdG9tRXJyb3IoRXhjZXB0aW9uKToKICAgIGRlZiBfX2luaXRfXyhzZWxmLCBtZXNzYWdlKToKICAgICAgICBzZWxmLm1lc3NhZ2UgPSBtZXNzYWdlCgojIExhbnphbmRvIGxhIGV4Y2VwY2nDs24gcGVyc29uYWxpemFkYQpyYWlzZSBDdXN0b21FcnJvcigiRXN0ZSBlcyB1biBlcnJvciBwZXJzb25hbGl6YWRvIik="))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'># Ejemplo: Creando una excepci&oacute;n personalizada
class CustomError(Exception):
    def __init__(self, message):
        self.message = message

# Lanzando la excepci&oacute;n personalizada
raise CustomError(&quot;Este es un error personalizado&quot;)</pre></div><div class='content'></div><h2>Uso de Excepciones Personalizadas</h2>
<div class='content'><p>Las excepciones personalizadas pueden ser usadas para proporcionar mensajes de error más específicos y para manejar errores de una manera más granular.</p>
</div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("IyBFamVtcGxvOiBVc2FuZG8gdW5hIGV4Y2VwY2nDs24gcGVyc29uYWxpemFkYSBlbiB1bmEgZnVuY2nDs24KY2xhc3MgTmVnYXRpdmVOdW1iZXJFcnJvcihFeGNlcHRpb24pOgogICAgcGFzcwoKZGVmIGNoZWNrX3Bvc2l0aXZlKG51bWJlcik6CiAgICBpZiBudW1iZXIgPCAwOgogICAgICAgIHJhaXNlIE5lZ2F0aXZlTnVtYmVyRXJyb3IoIk5vIHNlIHBlcm1pdGVuIG7Dum1lcm9zIG5lZ2F0aXZvcyIpCgp0cnk6CiAgICBjaGVja19wb3NpdGl2ZSgtNSkKZXhjZXB0IE5lZ2F0aXZlTnVtYmVyRXJyb3IgYXMgZToKICAgIHByaW50KGUp"))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'># Ejemplo: Usando una excepci&oacute;n personalizada en una funci&oacute;n
class NegativeNumberError(Exception):
    pass

def check_positive(number):
    if number &lt; 0:
        raise NegativeNumberError(&quot;No se permiten n&uacute;meros negativos&quot;)

try:
    check_positive(-5)
except NegativeNumberError as e:
    print(e)</pre></div><div class='content'></div><h1>Técnicas Avanzadas</h1>
<div class='content'></div><h2>Encadenamiento de Excepciones</h2>
<div class='content'><p>Puedes encadenar excepciones para proporcionar más contexto sobre el error. Esto se hace usando la palabra clave <code>from</code>.</p>
</div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("IyBFamVtcGxvOiBFbmNhZGVuYW1pZW50byBkZSBleGNlcGNpb25lcwp0cnk6CiAgICByYWlzZSBWYWx1ZUVycm9yKCJFcnJvciBpbmljaWFsIikKZXhjZXB0IFZhbHVlRXJyb3IgYXMgZToKICAgIHJhaXNlIEN1c3RvbUVycm9yKCJPY3VycmnDsyB1biBlcnJvciBwZXJzb25hbGl6YWRvIikgZnJvbSBl"))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'># Ejemplo: Encadenamiento de excepciones
try:
    raise ValueError(&quot;Error inicial&quot;)
except ValueError as e:
    raise CustomError(&quot;Ocurri&oacute; un error personalizado&quot;) from e</pre></div><div class='content'></div><h2>Relanzamiento de Excepciones</h2>
<div class='content'><p>A veces puede que quieras capturar una excepción, realizar algunas acciones y luego relanzar la misma excepción.</p>
</div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("IyBFamVtcGxvOiBSZWxhbnphbWllbnRvIGRlIHVuYSBleGNlcGNpw7NuCnRyeToKICAgIHJhaXNlIFZhbHVlRXJyb3IoIk9jdXJyacOzIHVuIGVycm9yIikKZXhjZXB0IFZhbHVlRXJyb3IgYXMgZToKICAgIHByaW50KCJNYW5lamFuZG8gZWwgZXJyb3IiKQogICAgcmFpc2UgICMgUmVsYW56YSBsYSBtaXNtYSBleGNlcGNpw7Nu"))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'># Ejemplo: Relanzamiento de una excepci&oacute;n
try:
    raise ValueError(&quot;Ocurri&oacute; un error&quot;)
except ValueError as e:
    print(&quot;Manejando el error&quot;)
    raise  # Relanza la misma excepci&oacute;n</pre></div><div class='content'></div><h1>Conclusión</h1>
<div class='content'><p>Lanzar excepciones es una parte fundamental del manejo de errores en Python. Al entender cómo lanzar excepciones incorporadas, crear excepciones personalizadas y usar técnicas avanzadas como el encadenamiento y el relanzamiento, puedes escribir código más robusto y mantenible. Recuerda siempre proporcionar mensajes de error significativos para ayudar con la depuración y manejar las excepciones de manera elegante para asegurar una buena experiencia de usuario.</p>
</div><div class='row navigation'>
	<div class='col-4'>
					<a href='handling-exceptions'>&#x25C4;Manejo de Excepciones</a>
			</div>
	<div class='col-4 text-center'>
		<a href="./" class="title">Lanzamiento de Excepciones</a>
	</div>
	<div class='col-4 text-end'>
					<a href='custom-exceptions'>Excepciones Personalizadas &#x25BA;</a>
			</div>
</div>
</div>
		</div>
		<div class="col-12 col-lg-4 publi" id="div_publi">
			<h1>Publicidad</h1>
			<p>Este espacio está destinado a publicidad.</p>
			<p>Si quieres ser patrocinador, contáctanos para incluir enlaces en esta zona: <a href='mailto:admin@campusempresa.cat'>admin@campusempresa.cat</a></p>
			<p>¡Gracias por colaborar!</p>
		</div>
	</div>
</div>

   <div class="container-xxl my-3">
	<div class="row">
		<div class="col">
			<footer>&copy; Copyright 2024. Todos los derechos reservados</footer>
		</div>
	</div>
</div>	

<div id="cookies_adv" style="display:none;">
	Fem servir galetes per millorar la teva experiència d'ús i oferir continguts adaptats als teus interessos
    <a href="#" id="btn_accept_cookies" class="button">Entès!</a>
    <a href="/cookies">Més informació</a>
</div>	

	</div>    
</body>
</html>
